### PR TITLE
Feature/31243/leeds crc refactor

### PR DESCRIPTION
**Pull request by shilpigoeldev**
## What?

I have refactored Leeds CRC importer with latest rules provided by Fiona , as per Planio ticket 35286

## Why?

Whole Leeds CRC had to be re-written again as older code wasn't getting the right denominators and counts.

## How?

Have re-written Leeds CRC importer as per rules sheet provided by Fiona.

## Testing?

QA signed off by Fiona after sharing the counts. Also added handler tests for each kind of case.

## Anything Else?
It still has rubocop errors for existing code in parent classes like genocolorectal.rb, which can't be edited in this scope of work as being used across all importers.

### CODE REVIEW

**Review by bshand — COMMENTED** [review]
I've requested a small performance-related change. Otherwise, this looks OK to me, though I don't really understand enough about the logic to easily follow the refactoring.

File: lib/import/colorectal/providers/leeds/leeds_handler_colorectal.rb [hunk 1]
```diff
@@ -8,641 +8,395 @@ module Leeds
...
+            else
+              7
+            end
+          end
+
+          def add_scope(genocolorectal, record)
+            genotype = record.raw_fields['genotype'].downcase.strip
+            mtype = record.raw_fields['moleculartestingtype'].downcase.strip
+            if genotype.match(/pms2\s-\smlpa\spred\snegative\sc5/i) ||
+               genotype.match(/pms2\s-\sconf\smlpa\spositive\sc5/i)
```
Comment (lauramccluskey1) [inline lib/import/colorectal/providers/leeds/leeds_handler_colorectal.rb:384]: I can't see this in the rules? Can you send me the updated ones please?

File: lib/import/colorectal/providers/leeds/leeds_handler_colorectal.rb [hunk 2]
```diff
@@ -1,648 +1,403 @@
...
+            elsif @report.include?('sequence variant c.1595G>A')
+              @teststatus = 2
+            end
+          end
+
+          def cleanup_report(report)
+            raw_report = report
+            exclude_statements = [
+              'Screening for mutations in MLH1, MSH2 and MSH6 is now in progress as requested.',
+              'MLPA and MSH2 analysis was not requested.',
```
Comment (lauramccluskey1) [inline lib/import/colorectal/providers/leeds/leeds_handler_colorectal.rb:329]: Would it be good to add these to the constants file instead?

File: lib/import/germline/genotype.rb [hunk 3]
```diff
@@ -336,7 +340,7 @@ def add_variant_impact(impact)
       end
 
       def add_variant_class(variant)
-        if variant.is_a?(Integer) && variant >= 1 && variant <= 5
+        if variant.is_a?(Integer) && variant >= 1 && variant <= 7
```
Comment (lauramccluskey1) [inline lib/import/germline/genotype.rb:343]: As this is an update to the assigning variant class in the genotype.rb file, have you checked that it won't affect any of the other importers?

Comment (shilpigoeldev) [reply lib/import/germline/genotype.rb:343]: Its extending the range that's allowed to be in variantpathclass  and adheres to Zvariantpathclass. If any other importer is able to extract c4/5 pathogenicity values and map upto 7 in future that should be fine.

File: lib/import/colorectal/providers/leeds/leeds_handler_colorectal.rb [hunk 4]
```diff
@@ -8,641 +8,395 @@ module Leeds
...
+            populate_genotype(record)
+          end
+
+          def populate_variables(record)
+            @geno = record.raw_fields['genotype']&.downcase
+            @report = cleanup_report(record.raw_fields['report'])
+            @moleculartestingtype = record.raw_fields['moleculartestingtype']&.downcase
+            @indicationcategory = record.raw_fields['indicationcategory']&.downcase
+            @genes_hash = YAML.safe_load(File.open(Rails.root.join(GENES_FILEPATH)))
+            @status_hash = YAML.safe_load(File.open(Rails.root.join(STATUS_FILEPATH)))
```
[the remark this answers is not in the document]
Comment (shilpigoeldev) [reply lib/import/colorectal/providers/leeds/leeds_handler_colorectal.rb]: Thanks @bshand , done in latest commit

File: lib/import/colorectal/providers/leeds/leeds_handler_colorectal.rb (hunk 1 above)
Comment (shilpigoeldev) [reply lib/import/colorectal/providers/leeds/leeds_handler_colorectal.rb:385]: Have mailed you the rules , and following - 

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/shilpigoel/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/shilpigoel/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">


Raw:genotype | Test Scope
-- | --
PMS2 - MLPA   pred negative C5 | Targeted
PMS2 - Conf   MLPA positive C5 | Targeted



</body>

</html>

Comment (lauramccluskey1) [reply lib/import/colorectal/providers/leeds/leeds_handler_colorectal.rb:385]: Thank you! This makes sense now

File: lib/import/colorectal/providers/leeds/leeds_handler_colorectal.rb [hunk 5]
```diff
@@ -1,648 +1,403 @@
...
+          def process_negative_genes(negative_genes, genocolorectal, genotypes)
+            negative_genes&.each do |gene|
+              genocolorectal_dup = genocolorectal.dup_colo
+              genocolorectal_dup.add_gene_colorectal(gene)
+              genocolorectal_dup.add_status(1)
+              genotypes << genocolorectal_dup
+            end
+          end
+
+          def allocate_genes(_record)
```
Comment (lauramccluskey1) [inline lib/import/colorectal/providers/leeds/leeds_handler_colorectal.rb:291]: record is not being used by the function- can this be removed?

File: lib/import/colorectal/providers/leeds/leeds_handler_colorectal.rb (hunk 2 above)
Comment (shilpigoeldev) [reply lib/import/colorectal/providers/leeds/leeds_handler_colorectal.rb]: Done in latest commit .

File: lib/import/colorectal/providers/leeds/leeds_handler_colorectal.rb (hunk 5 above)
Comment (shilpigoeldev) [reply lib/import/colorectal/providers/leeds/leeds_handler_colorectal.rb]: Done in latest commit

**Comment by lauramccluskey1** [comment]
Can you also add more comments please?

File: lib/import/colorectal/providers/leeds/leeds_handler_colorectal.rb [hunk 6]
```diff
@@ -1,696 +1,441 @@
...
 
-          # rubocop:disable Metrics/AbcSize
           def process_fields(record)
+            populate_variables(record)
+            return unless should_process
+
+            populate_and_persist_genotype(record)
+          end
+
+          # populate class variables that can be used over different methods
```
Comment (ollietulloch) [inline lib/import/colorectal/providers/leeds/leeds_handler_colorectal.rb:23]: I think these are instance variables, may be worth changing the comment

File: lib/import/colorectal/providers/leeds/leeds_handler_colorectal.rb [hunk 7]
```diff
@@ -1,696 +1,441 @@
...
+          def populate_variables(record)
+            @geno = record.raw_fields['genotype']&.downcase
+            @report = cleanup_report(record.raw_fields['report'])
+            @moleculartestingtype = record.raw_fields['moleculartestingtype']&.downcase
+            @indicationcategory = record.raw_fields['indicationcategory']&.downcase
+            @genes_panel = []
+          end
+
+          # checks if file has colorectal tests and should be processed under this importer
+          def should_process
```
Comment (ollietulloch) [inline lib/import/colorectal/providers/leeds/leeds_handler_colorectal.rb:33]: If this only returns true of false, could this be a predicate method?

File: lib/import/colorectal/providers/leeds/leeds_handler_colorectal.rb [hunk 8]
```diff
@@ -1,696 +1,441 @@
...
             res.map { |cur_genotype| @persister.integrate_and_store(cur_genotype) }
           end
-          # rubocop:enable Metrics/AbcSize
 
-          def add_organisationcode_testresult(genocolorectal)
-            genocolorectal.attribute_map['organisationcode_testresult'] = '699C0'
+          def process_variants_from_record(genocolorectal, record)
+            genotypes = []
+            allocate_genes
+            find_test_status(record)
```
Comment (ollietulloch) [inline lib/import/colorectal/providers/leeds/leeds_handler_colorectal.rb:62]: A small thing but naming methods consistently is really helpful regarding readability. `allocate_genes` and `find_test_status` do similar things, i.e. setting an instance variable, so perheps calling it `allocate_test_status` might be more consistent?

File: lib/import/colorectal/providers/leeds/leeds_handler_colorectal.rb [hunk 9]
```diff
@@ -1,696 +1,441 @@
...
+              STATUS_PANEL.each do |category, status|
+                @teststatus = status if @status_hash[category].include?(@geno)
+              end
+            end
+            # exceptional cases for 'mlh1 only -ve' genotype
+            exceptinal_teststatus if @geno == 'mlh1 only -ve'
+            @teststatus
+          end
+
+          def exceptinal_teststatus
```
Comment (ollietulloch) [inline lib/import/colorectal/providers/leeds/leeds_handler_colorectal.rb:326]: Should this be `exceptional_teststatus`?

File: lib/import/colorectal/providers/leeds/leeds_handler_colorectal.rb (hunk 6 above)
Comment (shilpigoeldev) [reply lib/import/colorectal/providers/leeds/leeds_handler_colorectal.rb]: Thanks @ollietulloch , done in latest commit .

File: lib/import/colorectal/providers/leeds/leeds_handler_colorectal.rb (hunk 8 above)
Comment (shilpigoeldev) [reply lib/import/colorectal/providers/leeds/leeds_handler_colorectal.rb]: Thanks @ollietulloch , done in latest commit .

File: lib/import/colorectal/providers/leeds/leeds_handler_colorectal.rb (hunk 9 above)
Comment (shilpigoeldev) [reply lib/import/colorectal/providers/leeds/leeds_handler_colorectal.rb]: Thanks @ollietulloch , done in latest commit .

**Review by ollietulloch — APPROVED** [review]

**Pull request merged**